lib/remote/statsreporter.cpp

```diff
@@ -0,0 +1,192 @@
+/******************************************************************************
+ * Icinga 2                                                                   *
+ * Copyright (C) 2012-2018 Icinga Development Team (https://www.icinga.com/)  *
+ *                                                                            *
+ * This program is free software; you can redistribute it and/or              *
+ * modify it under the terms of the GNU General Public License                *
+ * as published by the Free Software Foundation; either version 2             *
+ * of the License, or (at your option) any later version.                     *
+ *                                                                            *
+ * This program is distributed in the hope that it will be useful,            *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
+ * GNU General Public License for more details.                               *
+ *                                                                            *
+ * You should have received a copy of the GNU General Public License          *
+ * along with this program; if not, write to the Free Software Foundation     *
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.             *
+ ******************************************************************************/
+
+#include "base/application.hpp"
+#include "base/array.hpp"
+#include "base/configtype.hpp"
+#include "base/dictionary.hpp"
+#include "base/function.hpp"
+#include "base/objectlock.hpp"
+#include "base/scriptglobal.hpp"
+#include "base/statsfunction.hpp"
+#include "base/utility.hpp"
+#include "base/value.hpp"
+#include "remote/apifunction.hpp"
+#include "remote/endpoint.hpp"
+#include "remote/messageorigin.hpp"
+#include "remote/statsreporter.hpp"
+#include "remote/zone.hpp"
+#include <boost/bind.hpp>
+#include <boost/thread/mutex.hpp>
+#include <set>
+
+using namespace icinga;
+
+REGISTER_APIFUNCTION(ClusterStats, event, &StatsReporter::ClusterStatsAPIHandler);
+
+REGISTER_STATSFUNCTION(ClusterStats, &StatsReporter::StatsFunc);
+
+StatsReporter StatsReporter::m_Instance;
+
+StatsReporter::StatsReporter()
+{
+	Endpoint::OnConnected.connect(std::bind(&StatsReporter::OnConnected, this, _1));
+}
+
+void StatsReporter::OnConnected(const Endpoint::Ptr& endpoint)
+{
+	if (!m_HasBeenInitialized.test_and_set()) {
+		timer = new Timer;
+		timer->OnTimerExpired.connect(std::bind(&StatsReporter::ReportStats, this));
+		timer->SetInterval(10);
+		timer->Start();
+		timer->Reschedule(1);
+	}
+}
+
+void StatsReporter::ReportStats()
+{
+	Dictionary::Ptr message;
+
+	for (auto& zone : Zone::GetLocalZone()->GetAllParents()) {
```

Comment:
That will lead into many messages. When we talked about the possible algorithm here, I thought of sending the status to the immediate parent zone (and all endpoints), they update their local cache with what they receive from child zone, and then generate a new cluster message with aggregated data. That way the data may not be "live" but is always bound to "next hop, next hop".

Reply:
You're right and this was an #ElephantOversight.